### PR TITLE
Adds a new LanguageServer Builder API

### PR DESCRIPTION
**Pull request by david-driscoll**
This is a large... breaking change, but all existing tests pass.

* Replaces custom mediator framework with `MediatR`
* Adds support for dependency injection internally.
* Added some test helpers to make testing a little easier.
* Bumped target framework to `netstandard2.0`

### CODE REVIEW

**Comment by tintoy** [comment]
So is DI internal-only, or can consumers add their own stuff to the container?

Am just thinking it would be useful to register handlers that way if they have dependencies...

**Comment by david-driscoll** [comment]
That's on the list next, first stab was getting things wired up.  Next step is how best to add handlers.

**Comment by tintoy** [comment]
If you're open to a little bike-shedding, I might have some ideas around that (e.g. both Autofac's registration API and the ASP.NET. NET WebAPI controller registration behaviour have ergonomics it might be worth exploring...)

**Comment by david-driscoll** [comment]
Go right ahead!  🚲 🏠 away.

I just added (an untested part) that does pull the handlers from the container as well.

**Comment by tintoy** [comment]
I think as long as there is a way to manually control which handlers are registered (and when) then it'll be fine. I'll write up some thoughts this arvo and then post them here :)

**Comment by david-driscoll** [comment]
Those methods make sense.  Just wrapping a little bow around some of the pieces.

**Comment by david-driscoll** [comment]
Rebased.

File: sample/SampleServer/Program.cs [hunk 1]
```diff
@@ -28,10 +28,9 @@ static void Main(string[] args)
                     .WithLoggerFactory(new LoggerFactory())
                     .AddDefaultLoggingProvider()
                     .WithMinimumLogLevel(LogLevel.Trace)
+                    .WithHandler<TextDocumentHandler>()
```
Comment (tintoy) [inline sample/SampleServer/Program.cs]: BTW, looks to me like this overload will auto-register any JSON-RPC interfaces implemented by the handler type. Is that correct?

Comment (david-driscoll) [reply sample/SampleServer/Program.cs]: yeah, the handler collection will create a descriptor per `IJsonRpcHandler` interface it finds in the implementing class.

**Comment by david-driscoll** [comment]
Going to merge this for now, and test one of the pre-release packages against omnisharp.

**Comment by tintoy** [comment]
I'll look into adding similar support for the client this weekend :)

**Pull request merged**